cvmfs/catalog_mgr_client.cc

```diff
@@ -87,188 +88,281 @@ shash::Any ClientCatalogManager::GetRootHash() {
-                                     &cache_hash, signature_mgr_,
+  UniquePtr<CachedManifestEnsemble> ensemble(
+                       new CachedManifestEnsemble(fetcher_->cache_mgr(), this));
+  manifest_failure = manifest::Fetch("", repo_name_, local_newest_timestamp,
+                                     &local_newest_hash, signature_mgr_,
                                      fetcher_->download_mgr(),
-                                     &ensemble);
-  if (manifest_failure != manifest::kFailOk) {
-    LogCvmfs(kLogCache, kLogDebug, "failed to fetch manifest (%d - %s)",
-             manifest_failure, manifest::Code2Ascii(manifest_failure));
-
-    LoadError success_code = catalog::kLoadUp2Date;
-
-    // Network unavailable but cached copy updated externally?
-    std::map<PathString, shash::Any>::const_iterator iter =
-      mounted_catalogs_.find(mountpoint);
-    if (iter != mounted_catalogs_.end()) {
-      if (breadcrumb.IsValid() && (iter->second != cache_hash)) {
-        success_code = catalog::kLoadNew;
+                                     ensemble.weak_ref());
+
+  if (manifest_failure == manifest::kFailOk) {
+    // server has newest revision or no valid local revision
+    if (ensemble->manifest->revision() > local_newest_revision
+          || local_newest_revision == manifest::Breadcrumb::kInvalidRevision
+          // if revision is 0 both local and server, load catalog from server
+          // as local is most likely just "initialized" without valid value
+          || (ensemble->manifest->revision() == 0 &&
+              local_newest_revision == 0)) {
+      result->SetHash(ensemble->manifest->catalog_hash());
+      result->SetRootCtlgRevision(ensemble->manifest->revision());
+      result->SetRootCtlgLocation(kCtlgLocationServer);
+      if (ensemble->manifest->has_alt_catalog_path()) {
+        fixed_alt_root_catalog_ = true;
+      } else {
+        fixed_alt_root_catalog_ = false;
       }
-    }
 
-    if (catalog_path) {
-      LoadError success_code =
-        LoadCatalogCas(cache_hash, cvmfs_path, "", catalog_path);
-      if (success_code != catalog::kLoadNew)
-        return success_code;
-      loaded_catalogs_[mountpoint] = cache_hash;
+      result->TakeManifestEnsemble(
+                    static_cast<manifest::ManifestEnsemble*>(
+                                                           ensemble.Release()));
+      offline_mode_ = false;
+
+      return catalog::kLoadNew;
     }
+  }
+  LogCvmfs(kLogCache, kLogDebug,
+            "Failed fetch manifest from server: "
+            "manifest too old or server unreachable (%d - %s)",
+            manifest_failure, manifest::Code2Ascii(manifest_failure));
+
+  // corner case: server not reachable, local revision is 0. We can not
+  // distinguish if local revision number is valid or invalid. Most likely
+  // it is invalid - so fail
+  if (local_newest_revision == 0 && (manifest_failure != manifest::kFailOk)
```

Comment:
Does not need to be addressed here, but in general I think consistent parenthesis use would be preferable:
```suggestion
  if ((local_newest_revision == 0) && (manifest_failure != manifest::kFailOk)
```

Also: why do we need the revision check here? Isn't that strictly superseded by checking that the local hash is unset?

Reply:
ok lets try this and see what the cloudtests are saying.